File: SearchRace/src/main/java/com/codingame/game/Car.java
package com.codingame.game;

public class Car extends Unit {
    public double angle;
    public String message;
    public boolean debug;
    public int thrust = 0;

    public Point target;

    Car(double x, double y, double angle) {
        super(x, y);
        this.angle = angle;
        this.friction = Constants.CAR_FRICTION;
    }


    public void handleInput(String input, IPlayerManager manager) throws Exception{
        target = null;
        message = "";
        Car car = this;
        String[] splitted = input.split(" ");

        if(splitted[0].equals("EXPERT")){
            int angle = Integer.parseInt(splitted[1]);
            thrust = Integer.parseInt(splitted[2]);
            if(thrust < 0 || thrust > Constants.CAR_MAX_THRUST) {
                manager.addGameSummary("Invalid thrust. Please keep between 0 and " + Constants.CAR_MAX_THRUST);
                throw new Exception( "Invalid thrust");
            }

            if(angle < -18 || angle > 18){
                manager.addGameSummary("Invalid angle. Please keep between -18 and 18.");
                throw new Exception("Invalid angle");
            }

            car.handleExpertInput(angle, thrust);
            if(splitted.length > 3){
                int totalLength = ("EXPERT " +angle+" "+thrust+" ").length();
                car.message = input.substring(totalLength);

                if(car.message.length() > 20){
                    car.message = car.message.substring(0, 20);
                }
            }else{
                car.message="";
            }

        }
        else{
            int x = Integer.parseInt(splitted[0]);
            int y = Integer.parseInt(splitted[1]);
            target = new Point(x, y);
            thrust = Integer.parseInt(splitted[2]);
            if(thrust < 0 || thrust > Constants.CAR_MAX_THRUST) {
                manager.addGameSummary("Invalid thrust. Please keep between 0 and " + Constants.CAR_MAX_THRUST);
                throw new Exception( "Invalid thrust");
            }

            car.handleInput(x, y, thrust);
            if(splitted.length > 3){
                int totalLength = (x+" "+y+" "+thrust+" ").length();
                car.message = input.substring(totalLength);
                if(car.message.length() > 20){
                    car.message = car.message.substring(0, 20);
                }
            }else{
                car.message="";
            }
        }

        if(car.message.contains("debug")){
            debug = true;
        }
    }

    public void handleExpertInput(int angle, int thrust){
        double newAngle = Math.toDegrees(this.angle) + angle;
        this.angle = Math.toRadians(newAngle);
        thrustTowardsHeading(thrust);
    }

    public void handleInput(int x, int y, int thrust){
        if (this.x != x || this.y != y) {
            double angle = this.getAngle(new Point(x, y));
            double relativeAngle = shortAngleDist(this.angle, angle);
            if (Math.abs(relativeAngle) >= Constants.MAX_ROTATION_PER_TURN) {
                angle = this.angle + Constants.MAX_ROTATION_PER_TURN * Math.signum(relativeAngle);
            }

            this.angle = angle;
            thrustTowardsHeading(thrust);
        }
    }


    private void thrustTowardsHeading(int thrust){
        double vx = Math.cos(angle) * thrust;
        double vy = Math.sin(angle) * thrust;

        this.vx += vx;
        this.vy += vy;
    }

    @Override
    public void adjust() {
        super.adjust();
        double degrees = Math.round(Math.toDegrees(angle));
        this.angle = Math.toRadians(degrees);
        while(this.angle > Math.PI*2) this.angle-= Math.PI*2;
        while(this.angle < 0)this.angle+= Math.PI*2;
    }

    private static double shortAngleDist(double a0, double a1) {
        double max = Math.PI * 2;
        double da = (a1 - a0) % max;
        return 2 * da % max - da;
    }
}

File: CGSearchRace/Car.py
from math import cos, sin, radians, degrees, copysign

from CGSearchRace.Constants import Constants
from CGSearchRace.Point import Point
from CGSearchRace.Unit import Unit


class Car(Unit):

    def __init__(self, x, y, angle):
        super().__init__(x, y)
        self.angle = angle
        self.friction = Constants.CAR_FRICTION
        self.thrust = 0
        self.prev_angle = 0
        self.target = None
        self.debug = False

    def handle_input_(self, user_input):
        self.prev_angle = self.angle
        self.target = None
        split = user_input.split(" ")

        if split[0] == "EXPERT":
            try:
                angle = int(split[1])
                thrust = int(split[2])
            except Exception:
                raise Exception("Can't parse expert input integers. ")

            if thrust < 0 or thrust > Constants.CAR_MAX_THRUST:
                raise Exception("Invalid thrust")

            if angle < -18 or angle > 18:
                raise Exception("Invalid angle")

            self.handle_expert_input(angle, thrust)
            if split.length > 3:
                total_length = len(f"EXPERT :angle :thrust")
                self.message = user_input.substring(total_length)
            else:
                self.message = ""
        else:
            try:
                x = int(split[0])
                y = int(split[1])
                self.target = Point(x, y)
                thrust = int(split[2])
            except Exception:
                raise Exception("Can't parse action. Has to be X Y THRUST.")

            if thrust < 0 or thrust > Constants.CAR_MAX_THRUST:
                raise Exception("Invalid thrust")

            self.handle_input(x, y, thrust)
            if len(split) > 3:
                total_length = len(f":x :y :thrust")
                self.message = user_input.substring(total_length)
            else:
                self.message = ""

        if "debug" in self.message:
            self.debug = True

    def handle_expert_input(self, angle: int, thrust: int):
        new_angle = degrees(self.angle) + angle
        self.angle = radians(new_angle)
        self.thrust_towards_heading(thrust)

    def handle_input(self, x: int, y: int, thrust: int):
        if self.x != x or self.y != y:
            angle = self.get_angle(Point(x, y))
            relative_angle = self.short_angle_dist(self.angle, angle)
            if abs(relative_angle) >= Constants.MAX_ROTATION_PER_TURN:
                angle = self.angle + Constants.MAX_ROTATION_PER_TURN * copysign(1, relative_angle)

            self.angle = angle
            self.thrust_towards_heading(thrust)

    def thrust_towards_heading(self, thrust: int):
        vx = cos(self.angle) * thrust
        vy = sin(self.angle) * thrust
        self.vx += vx
        self.vy += vy

    def adjust(self):
        super().adjust()
        self.angle = radians(round(degrees(self.angle)))
        self.angle = self.angle % (Constants.PI * 2)

    def short_angle_dist(self, a0: float, a1: float):
        max_angle = Constants.PI * 2
        da = (a1 - a0) % max_angle
        return 2 * da % max_angle - da
